solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -95,10 +103,250 @@ pub mod audius_data {
+        let mut eth_addresses = BTreeMap::new();
+        if !proposers.iter().all(move |(proposer, authority)| return match eth_addresses.insert(proposer.owner_eth_address, true) {
+            Some(_) => false,
+            None => proposer.authority == authority.key()
+        }) {
+            // duplicate owner eth address - err
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Confirm that the base used for user account seed is derived from this Audius admin storage account
+        let (derived_base, _) = Pubkey::find_program_address(
+            &[&ctx.accounts.admin.key().to_bytes()[..32]],
+            ctx.program_id,
+        );
+
+        if derived_base != base {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Refer to context here - https://docs.solana.com/developing/programming-model/transactions#multiple-instructions-in-a-single-transaction
+        let dummy_owner_field = Pubkey::from_str("11111111111111111111111111111111").unwrap();
+
+        let content_node = &mut ctx.accounts.content_node;
+        content_node.authority = dummy_owner_field;
+
+        Ok(())
+    }
+
+    /// Update a user's replica set
+    pub fn update_user_replica_set(
+        ctx: Context<UpdateUserReplicaSet>,
+        base: Pubkey,
+        _user_handle: UserHandle,
+        ursm: [u16;3],
+        _ursm_bumps: [u8; 3]
+    ) -> Result<()> {
+        let user_ursm = &ctx.accounts.user.ursm;
+
+        let did_user_sign = ctx.accounts.user.authority == ctx.accounts.cn_authority.key();
+        let mut is_valid_authority = did_user_sign;
+        let proposers = [&ctx.accounts.cn1, &ctx.accounts.cn2, &ctx.accounts.cn3];
+
+        // Validate that a content node in the replica signed the tx if the user's authority did not
+        if !is_valid_authority {
+            for proposer in proposers {
+                if proposer.authority == ctx.accounts.cn_authority.key() {
+                    is_valid_authority = true;
+                    if !user_ursm.iter().any(|cn| {
+                        let (base_pda, _bump) = Pubkey::find_program_address(
+                            &[
+                                &base.to_bytes()[..32],
+                                CONTENT_NODE_SEED_PREFIX,
+                                &cn.to_le_bytes()
+                            ], &ctx.program_id);
+                        return base_pda == proposer.key();
+                    }) {
+                        msg!("Audius::UpdateUserReplicaSet bad");
+                        return Err(ErrorCode::Unauthorized.into());
+                    }
+                }
+            }    
+        }
+
+        if !is_valid_authority {
+            msg!("Audius::UpdateUserReplicaSet oterh");
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Confirm that the base used for user account seed is derived from this Audius admin storage account
```

Comment:
maybe this should be checked first?

Reply:
I think it's fine either way because the tx will fail in the transaction simulation - so it will not waste SOL - but we should be consistent - moving to top